### ISSUE TITLE
`dataclass_transform` only respects parent class' field annotations

### ISSUE THREAD

**Issue by BlueGlassBlock** [microsoft/pyright#3907]
`dataclass_transform` only respects parent class' field annotations, not `__init__` signature.

### Code
```py
@dataclass_transform(kw_only_default=True)
class M(type):
    ...


class A(metaclass=M):
    a: str

    def __init__(self, a: str):
        self.a = a


class B(A):
    ...

# B's __init__ signature inferred as: (*, a: str) -> None
B("sample") # Expected 0 positional arguments.
```

### Expected behaviour
The code passes type checking, because `B`'s `__init__` method is overridden in `A`.

### Additional information
`mypy` and `pyre` both let the code pass type checking.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
I think I will need to revert this change. It breaks pydantic, which appears to assume pyright's previous behavior. It defines `BaseModel` as follows:

```python
@__dataclass_transform__(kw_only_default=True, field_descriptors=(Field, FieldInfo))
class ModelMetaclass(ABCMeta): ...

class BaseModel(Representation, metaclass=ModelMetaclass):
    def __init__(__pydantic_self__, **data: Any) -> None: ...
```

With the change requested in this bug report, any model that derives from `BaseModel` fails to get the intended synthesized `__init__` method.

**Comment by erictraut** [comment]
I've reverted the change until we have a chance to discuss this further. 

If mypy and pyre behave differently, I would think that they would not work correctly with pydantic either. Although IIRC they both have special-case logic to handle pydantic, which might explain why they're not hitting this issue.

Perhaps we need to update PEP 681 to clarify the behavior here and then get everyone (libraries and/or type checkers) to all conform to that behavior. cc @debonte

**Comment by debonte** [comment]
Sorry this is a bit long, but I wanted to document my thought process for "future me". If you want to jump to the punch line, go to the ["Behavior in this specific case"](#anchor) section.

### Background (`dataclass_transform` and class hierarchies)
PEP 681 [states](https://peps.python.org/pep-0681/#dataclass-semantics):

> Except where stated otherwise in this PEP, classes impacted by dataclass_transform...are assumed to behave like stdlib dataclass.

Therefore, generally when I'm confused about how `dataclass_transform` should behave, I look at what `dataclass` does. Of course, you can only create a stdlib dataclass via the `@dataclass` decorator, not through inheritance. But still, I think the core question is: "What would the equivalent `@dataclass`-based code look like and how does it behave?"

I see two options for the `@dataclass` equivalent of `A` inheriting from `M`:
1. Only `A` is annotated with `@dataclass`.
```python
@dataclass(kw_only=True)
class A():
    a: str

    def __init__(self, a: str):
        self.a = a

class B(A):
    ...
```
2. Both `A` and `B` are annotated with `@dataclass`.
```python
@dataclass(kw_only=True)
class A():
    a: str

    def __init__(self, a: str):
        self.a = a

@dataclass
class B(A):
    ...
```

If we choose option 1, we do indeed get the behavior that @BlueGlassBlock is requesting. However, that is because `B` is not a `dataclass`. With option 1, if we want `B` to have dataclass-like behaviors, we'd need to use multiple inheritance (`B` inherits from both `A` and `M`), which feels very awkward. 

I would argue that option 2 is better -- it feels more natural for all classes derived from `A` to have dataclass-like behaviors. And I believe that was our assumption when writing the PEP. There are a couple of spots that imply this:

- The ["The dataclass_transform decorator"](https://peps.python.org/pep-0681/#the-dataclass-transform-decorator) section says the following. Unfortunately, I can see that this could perhaps be interpreted as only referring to classes that derive **directly** from the decorated class:
> If `dataclass_transform` is applied to a class, dataclass-like semantics will be assumed for any class that derives from the decorated class or uses the decorated class as a metaclass.
- The `Vehicle`/`Car` examples in the ["Dataclass semantics"](https://peps.python.org/pep-0681/#dataclass-semantics) section assume this behavior.

So, I believe option 2 is correct/expected. If a library using `dataclass_transform` does not like this behavior and wants to allow only some (not all) classes in a class hierarchy to have dataclass-like behaviors, that library can expose a decorator instead of (or in addition to) a base class/metaclass.

<a id="anchor" />

### Behavior in this specific case

So, based on the above, here's the full `@dataclass`-equivalent of @BlueGlassBlock's code:

```python
from dataclasses import dataclass

@dataclass(kw_only=True)
class A():
    a: str

    def __init__(self, a: str):
        self.a = a
        
@dataclass
class B(A):
    ...

B("sample")
```

And in this case `dataclass` generates an `__init__` method for `B`, in which the `a` parameter is keyword-only: `__init__(self, *, a: str) -> None`

So I believe that Pyright's current behavior is correct, and that Pydantic's current use of `dataclass_transform` is proper.

@erictraut, do you agree?

### Action items
If we're in agreement, I'll ask Jelle what the best approach to documenting/clarifying this would be. If we can make the clarification in the PEP, then I would suggest two changes:
1. Add another level of inheritance in the class and metaclass examples [here](https://peps.python.org/pep-0681/#class-example).
2. Clarify that the following statement in ["Dataclass semantics"](https://peps.python.org/pep-0681/#dataclass-semantics) is about specific classes, not class hierarchies. That is, method synthesis occurs unless the specific class in question has an explicit declaration of that method -- base class declarations are not considered.
>As with dataclass, method synthesis is skipped if it would overwrite a method that is explicitly declared within the class. For example, if a class declares an `__init__` method explicitly, an `__init__` method will not be synthesized for that class.

**Comment by erictraut** [comment]
Thanks for investigating this @debonte. I agree with your analysis, conclusions, and recommended action items.

**Comment by debonte** [comment]
@JelleZijlstra, based on this issue there's some confusion about how class hierarchies deriving from a dataclass_transform-decorated class/metaclass work. Can I change PEP 681 as described [above](#actions)? If not, what would be the best way to document this behavior? It's really a clarification, not a change in behavior.

**Comment by JelleZijlstra** [comment]
I think it's fine to clarify this in the PEP.

In the longer term I'd like there to be a living typing spec that replaces the various PEPs. Then clarifications like this can simply be updates to the PEP.

**Comment by BlueGlassBlock** [comment]
Thanks for clarifying this. Complying with the `dataclasses` module seems to be reasonable after all.
